src/server/pfs/persist/driver.go

```diff
@@ -0,0 +1,462 @@
+		blockClient:  pfs.NewBlockAPIClient(clientConn),
+		dbAddress:    dbAddress,
+		dbName:       dbName,
+		dbClient:     dbClient,
+	}, nil
+}
+
+func InitDB(address string, databaseName string) error {
+	session, err := dbConnect(address)
+	if err != nil {
+		return err
+	}
+	defer session.Close()
+
+	// Create the database
+	if _, err := gorethink.DBCreate(databaseName).RunWrite(session); err != nil {
+		return err
+	}
+
+	// Create tables
+	for _, table := range tables {
+		tableCreateOpts := tableToTableCreateOpts[table]
+		if _, err := gorethink.DB(databaseName).TableCreate(table, tableCreateOpts...).RunWrite(session); err != nil {
+			return err
+		}
+	}
+
+	// Create indexes
+	if _, err := gorethink.DB(databaseName).Table(commitTable).IndexCreateFunc(commitBranchIndex, func(row gorethink.Term) interface{} {
+		return row.Field("BranchClocks").Map(func(branchClock gorethink.Term) interface{} {
+			lastClock := branchClock.Field("Clocks").Nth(-1)
+			return []interface{}{
+				lastClock.Field("Branch"),
+				lastClock.Field("Clock"),
+			}
+		})
+	}, gorethink.IndexCreateOpts{
+		Multi: true,
+	}).RunWrite(session); err != nil {
+		return err
+	}
+
+	// Wait for indexes to be ready
+	if _, err := gorethink.DB(databaseName).Table(commitTable).IndexWait(commitBranchIndex).RunWrite(session); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func RemoveDB(address string, databaseName string) error {
+	session, err := dbConnect(address)
+	if err != nil {
+		return err
+	}
+	defer session.Close()
+
+	// Create the database
+	if _, err := gorethink.DBDrop(databaseName).RunWrite(session); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func dbConnect(address string) (*gorethink.Session, error) {
+	return gorethink.Connect(gorethink.ConnectOpts{
+		Address: address,
+		Timeout: connectTimeoutSeconds * time.Second,
+	})
+}
+
+func validateRepoName(name string) error {
+	match, _ := regexp.MatchString("^[a-zA-Z0-9_]+$", name)
+
+	if !match {
+		return fmt.Errorf("repo name (%v) invalid: only alphanumeric and underscore characters allowed", name)
+	}
+
+	return nil
+}
+
+func (d *driver) getTerm(table Table) gorethink.Term {
+	return gorethink.DB(d.dbName).Table(table)
+}
+
+func (d *driver) CreateRepo(repo *pfs.Repo, created *google_protobuf.Timestamp,
+	provenance []*pfs.Repo, shards map[uint64]bool) error {
+
+	err := validateRepoName(repo.Name)
+	if err != nil {
+		return err
+	}
+
+	_, err = d.getTerm(repoTable).Insert(&Repo{
+		Name:    repo.Name,
+		Created: created,
+	}).RunWrite(d.dbClient)
+	return err
+}
+
+func (d *driver) InspectRepo(repo *pfs.Repo, shards map[uint64]bool) (repoInfo *pfs.RepoInfo, retErr error) {
+	cursor, err := d.getTerm(repoTable).Get(repo.Name).Default(gorethink.Error("value not found")).Run(d.dbClient)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err := cursor.Close(); err != nil && retErr == nil {
+			retErr = err
+		}
+	}()
+	rawRepo := &Repo{}
+	cursor.Next(rawRepo)
```

Comment:
There's a function on cursors called [`One`](https://godoc.org/github.com/dancannon/gorethink#Cursor.One) which reads out a single value, checks for errors and closes the cursor. Would make this function a little bit nicer.